practical_assessment/master_controller/scripts/object_navigation.py

```diff
@@ -45,111 +62,152 @@ def __init__(self):
+        self.curr_orientation = yaw
 
     def object_control_cb(self, action):
         self.object_action = action.data
         if self.object_action == START:
-            if not self.is_goal_sent:
-                self.navigate_to_object()
+            self.navigate_to_object()
         elif self.object_action == STOP:
-            if self.is_goal_sent:
-                rospy.loginfo("Stopping object navigation...")
-                self.ac.cancel_goal()
-            self.is_goal_sent = False
+            rospy.loginfo("Stopping object navigation...")
+            self.move_cmd(Twist())
+            self.is_moving = False
         elif self.object_action == REACHED:
             rospy.loginfo("Object reached!")
         else:
             rospy.loginfo("Invalid command. Waiting...")
 
     def navigate_to_object(self):
         '''
-        Checks if object type and pose is received and starts
+        Checks if object name and pose is received and starts
         navigating to object
         '''
-        rospy.loginfo("Starting object navigation...")
-        if not self.object_type:
-            rospy.logerr("Object type is not received. Waiting...")
+        if not self.is_moving:
+            rospy.loginfo("Starting object navigation...")
+
+        if not self.object_name:
+            rospy.logerr("Object name is not received. Waiting...")
             return
         elif not self.object_pose:
             rospy.logerr("Estimated object pose is not received. Waiting...")
             return
 
-        rospy.loginfo("Navigation to object {}. Estimated position at ({}, {})"
-                      .format(self.object_type,
-                              self.object_pose.x,
-                              self.object_pose.y))
-        self.move_to_goal()
+        if not self.is_moving:
+            rospy.loginfo("Navigation to object {}. Estimated position at ({}, {})"
+                          .format(self.object_name,
+                                  self.object_pose.x,
+                                  self.object_pose.y))
+        self.calculate_movement()
 
-    def move_to_goal(self):
+    def calculate_movement(self):
         '''
-        Calculate and move to the point 0.5m away from object
+        Calculates and moves to the point 0.5m away from object
         on the line between object and robot
         '''
         if not self.curr_position:
+            rospy.loginfo("Current position from /amcl is not received")
             return
 
-        # Current distance between robot and object
-        curr_dist = sqrt((self.object_pose.x - self.curr_position.x) ** 2 +
-                         (self.object_pose.y - self.curr_position.y) ** 2)
+        # calculation of target point should be done only once
+        if not self.is_moving:
+            # initial distance between robot and object
+            initial_dist = sqrt((self.object_pose.x - self.curr_position.x) ** 2 +
```

Comment:
Im confused about this, is this meant to be the actual distance to the object (are you using lidar somewhere) or distance from the center of the screen?

Reply:
I thought I'm getting the coordinates of the object from DetectedObject msg. Reimplemented object navigation using lidar

Reply:
So I think what you need to do is try to get the x coord in the middle of the screen (straight ahead) and then drive forward until lidar in front says <0.5m. Then you can announce you're at the object